lib/spack/spack/ci.py

```diff
@@ -514,6 +515,63 @@ def format_job_needs(phase_name, strip_compilers, dep_jobs,
     return needs_list
 
 
+def get_change_revisions():
+    """If this is a git repo get the revisions to use when checking
+    for changed packages and spack core modules."""
+    git_dir = os.path.join(spack.paths.prefix, '.git')
+    if os.path.exists(git_dir) and os.path.isdir(git_dir):
+        # TODO: This will only find changed packages from the last
+        # TODO: commit.  While this may work for single merge commits
+        # TODO: when merging the topic branch into the base, it will
+        # TODO: require more thought outside of that narrow case.
```

Comment:
Perhaps `git merge-base HEAD develop`? Better would be to have that be default and allow passing it in, then use the PR target branch in CI, that'll always get the whole thing.

Reply:
possibly better b/c you can still do a shallow clone: use the GitHub API.  See https://github.com/spack/spackbot/blob/main/spackbot/handlers/labels.py#L137-L175

See docs here: https://developer.github.com/v3/pulls/#list-pull-requests-files

Reply:
Can we get another identity to use for this api requests?  I ask because GitHub is already rate-limiting the one we have for some part of most hours.

Otherwise, yeah, we can do a full clone and get the merge-base between the parents of the merge commit we're testing.  Another suggestion (thanks to @opadron) was we just set the fetch depth to some "reasonable" value, and if we can't find the merge-base for PRs, print a message suggesting rebase.

I'm doing some experiments with all the branches I cloned from gitlab yesterday to see what "reasonable" could look like for that.

Reply:
Actions have their own GITHUB_TOKEN that you can use for API calls. It won’t have permission to write to the repo or to add users to teams like spackbot needs but it can query things like files modified in a PR.

Reply:
I guess since this is running in GitLab not GitHub I’m not sure you actually get that token though… hm.

Reply:
Here we go — you can do it with the GitLab API: https://docs.gitlab.com/ee/api/repositories.html#compare-branches-tags-or-commits

Reply:
For PR testing where we test the merge commit from github, my opinion is we want the values in this PR now (compare `HEAD^` and `HEAD`).  I wrote a little python script to check all the branches I pulled yesterday, and compare the changed files using the current approach against the changed files as returned by the github api.  Of the 498 branches my script checked, there were 14 where the file list didn't match and I looked at those PRs on github and found that they all fell into one of the following categories:

- waiting for style check to pass
- conflicts and cannot be merged
- draft PR
- branch pushed again between when I cloned the repo and ran the script

The first three conditions are noticed by the sync script and it avoids pushing in those cases.

This makes sense to me since I'd expect the merge commit made by github and returned to us in the gh api request for a PR to be consistent with the gh api request for the changed files of that same PR.

Reply:
But I agree with @trws that it would be better to have a reasonable default and allow passing it in.

Reply:
If we do it that way, different CI providers can provide it however makes sense.  In an action use the GH api, in GL use the GL api or something else.  Another option for dealing with the slow checkouts, that I was looking at for the other tests because many (**MANY**) unit tests fail without a significant chunk of a clone for some reason, is to switch to using `--shallow-since=year/mm/dd` with the day before the PR began.  It's guaranteed to have everything the PR touches, uses only the git API, and cuts down the spack repo for a PR like this one to only ~10mb.

```console
❯ g clone --shallow-since=2022/03/22 https://github.com/spack/spack spack-shallow
Cloning into 'spack-shallow'...
remote: Enumerating objects: 16406, done.
remote: Counting objects: 100% (16406/16406), done.
remote: Compressing objects: 100% (9358/9358), done.
remote: Total 16406 (delta 1293), reused 9904 (delta 959), pack-reused 0
Receiving objects: 100% (16406/16406), 9.79 MiB | 13.84 MiB/s, done.
Resolving deltas: 100% (1293/1293), done.
```

The downside is that it may have to be cloned from upstream, then fetch in the changes from a fork, the forks don't always allow that kind of shallow query.

Reply:
This is really my last concern, there are a good number of longer commit chains in PRs, is the plan to leave this TODO in place?

Reply:
If your concern is about whether this is going to work for gitlab pipelines run on spack PRs, I took some steps (described [here](https://github.com/spack/spack/pull/29697#discussion_r834912192), [here](https://github.com/spack/spack/pull/29697#issuecomment-1078604570), and [here](https://github.com/spack/spack/pull/29697#issuecomment-1079112231)) to make sure this approach is going to be sufficient for our intended usage.

You mentioned the ability to do things differently if we're in a github action, but we're relatively far from being able to generate a pipeline that runs in github actions (or any other CI provider), so I'm hesitant to hold this up over that.  

I could do something super-quick like allow you to override those values with a couple of environment variables.  Would that be useful for something you have mind (and get you on board for merging this :laughing:)?  I'd be happy to do that, and I'd be really interested what it is you have in mind.

Also, I got the feeling @tgamblin wanted to reduce the number of jobs we're generating the past couple weeks, and that time was of the essence (I guess so, since he just merged it).

Reply:
Nah, if it's getting the difference between the PR and develop (however it's getting it) I'm good with that.  The only thing I think would really be practical if we need the state back to merge base and still want shallow clones is to use the time-based shallow clone option.  If we don't need that to get the list of everything changed between the PR and develop then I wouldn't worry about it at all at this point.

Reply:
My understanding is we do not need it because we clone with depth 2 from the merge commit that's built for each PR.  On PR branches, that is all changes on the PR.  If we were to use this on `develop` or on some branch, yeah it'd just look at the difference with the last commit, but IMO that's the only thing that could make sense with pruning -- look at first parent.

But we wouldn't use pruning on `develop` or some branch -- we'll be using the old method there, where we just rebuild what isn't in cache.  Since the goal on a branch like `develop` is really to populate the cache, I don't see that as a problem. It's on the PRs where we get redundant rebuilds and want to reduce the builds.

Reply:
Yup, in case my wording was unclear I completely agree with that.  As long as
the PR builds are getting all the changes in the PR I don't see a problem.
Still might be nice to be able to specify it for testing outside automated CI,
but that's by no means pressing.